pkg/kubelet/pod_workers.go

```diff
@@ -138,57 +135,61 @@ type podWorkers struct {
-			// This is a blocking call that would return only if the cache
-			// has an entry for the pod that is newer than minRuntimeCache
-			// Time. This ensures the worker doesn't start syncing until
-			// after the cache is at least newer than the finished time of
-			// the previous sync.
-			status, err := p.podCache.GetNewerThan(podUID, lastSyncTime)
-			if err != nil {
+
+	for {
+		select {
+		case update := <-podUpdates:
+			err := func() error {
+				// This is a blocking call that would return only if the cache
+				// has an entry for the pod that is newer than minRuntimeCache
+				// Time. This ensures the worker doesn't start syncing until
+				// after the cache is at least newer than the finished time of
+				// the previous sync.
+				status, err := p.podCache.GetNewerThan(uid, lastSyncTime)
+				if err != nil {
+					return err
+				}
+				err = p.syncPodFn(syncPodOptions{
+					mirrorPod:      update.MirrorPod,
+					pod:            update.Pod,
+					podStatus:      status,
+					killPodOptions: update.KillPodOptions,
+					updateType:     update.UpdateType,
+				})
+				lastSyncTime = time.Now()
 				return err
+			}()
+			// notify the call-back function if the operation succeeded or not
+			if update.OnCompleteFunc != nil {
+				update.OnCompleteFunc(err)
 			}
-			err = p.syncPodFn(syncPodOptions{
-				mirrorPod:      update.MirrorPod,
-				pod:            update.Pod,
-				podStatus:      status,
-				killPodOptions: update.KillPodOptions,
-				updateType:     update.UpdateType,
-			})
-			lastSyncTime = time.Now()
-			return err
-		}()
-		// notify the call-back function if the operation succeeded or not
-		if update.OnCompleteFunc != nil {
-			update.OnCompleteFunc(err)
-		}
-		if err != nil {
-			glog.Errorf("Error syncing pod %s (%q), skipping: %v", update.Pod.UID, format.Pod(update.Pod), err)
-			// if we failed sync, we throw more specific events for why it happened.
-			// as a result, i question the value of this event.
-			// TODO: determine if we can remove this in a future release.
-			// do not include descriptive text that can vary on why it failed so in a pathological
-			// scenario, kubelet does not create enough discrete events that miss default aggregation
-			// window.
-			p.recorder.Eventf(update.Pod, v1.EventTypeWarning, events.FailedSync, "Error syncing pod")
+			if err != nil {
+				glog.Errorf("Error syncing pod %s (%q), skipping: %v", uid, format.Pod(update.Pod), err)
+				// if we failed sync, we throw more specific events for why it happened.
+				// as a result, i question the value of this event.
+				// TODO: determine if we can remove this in a future release.
+				// do not include descriptive text that can vary on why it failed so in a pathological
+				// scenario, kubelet does not create enough discrete events that miss default aggregation
+				// window.
+				p.recorder.Eventf(update.Pod, v1.EventTypeWarning, events.FailedSync, "Error syncing pod")
+			}
+			p.wrapUp(uid, err)
+		default:
```

Comment:
Are you sure you want this to busy loop any time we're waiting for an update to appear?

Reply:
Actually I don't want to do this, but I don't think up a good idea how to check a channel is empty.Could you give me a suggestion?

Reply:
Now that you've removed `lastUndeliveredWorkUpdate`, is `p.isWorking` used anywhere important? You may be able to just skip `p.finishWork(uid)` (which only sets `p.isWorking[uid] = false`) and go back to the `for update := range podUpdates` pattern.